File: backends/java-spring/src/test/java/com/authorization/integration/AuthIntegrationTest.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.integration;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;

import com.authorization.Application;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.http.client.JdkClientHttpRequestFactory;
import org.springframework.test.web.servlet.client.EntityExchangeResult;
import org.springframework.test.web.servlet.client.RestTestClient;

/**
 * 認証（Auth）APIの実DB/Redis接続統合テストです。
 * kotlin-ktorの{@code integration/AuthIntegrationTest.kt}に相当します。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
@SpringBootTest(classes = Application.class, webEnvironment = RANDOM_PORT)
class AuthIntegrationTest {

    @LocalServerPort
    private int port;

    private RestTestClient client;

    @BeforeEach
    void setUp() {
        TestHelper.truncateTables();
        client = RestTestClient.bindToServer().baseUrl("http://localhost:" + port).build();
    }

    @Test
    void meReturnsProfileWhenAuthenticated() {
        var staff = TestHelper.createStaff();

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/me")
                .header("Cookie", "staff_id=" + staff.id())
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
        assertThat(((Number) result.getResponseBody().get("staff_id")).longValue()).isEqualTo(staff.id());
    }

    @Test
    void meReturns401WhenUnauthenticated() {
        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/me")
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(401);
    }

    @Test
    void loginReturnsInfoWhenAuthenticated() {
        var staff = TestHelper.createStaff();

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/login")
                .header("Cookie", "staff_id=" + staff.id())
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
        // StaffVoはid（staff_idではない）フィールドをそのままJSON化するため、キー名が/meと異なる。
        assertThat(((Number) result.getResponseBody().get("id")).longValue()).isEqualTo(staff.id());
    }

    @Test
    void loginReturns401WhenUnauthenticated() {
        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/login")
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(401);
    }

    @Test
    void logoutReturns200() {
        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/logout")
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
    }

    private RestTestClient noRedirectClient() {
        var jdk = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).build();
        return RestTestClient.bindToServer(new JdkClientHttpRequestFactory(jdk))
                .baseUrl("http://localhost:" + port)
                .build();
    }

    @Test
    void githubRedirectIssuesNonceCookieAndEmbedsItInState() {
        EntityExchangeResult<byte[]> result = noRedirectClient().get()
                .uri("/auth/github/redirect?token=inv-token")
                .exchange()
                .expectBody()
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(302);
        List<String> cookies = result.getResponseHeaders().get("Set-Cookie");
        String oauthCookie = cookies.stream().filter(c -> c.startsWith("oauth_state=")).findFirst().orElseThrow();
        Matcher m = Pattern.compile("oauth_state=([0-9a-f]+)").matcher(oauthCookie);
        assertThat(m.find()).isTrue();
        assertThat(oauthCookie).contains("HttpOnly");
        String expected = URLEncoder.encode("java|" + m.group(1) + "|inv-token", StandardCharsets.UTF_8);
        assertThat(result.getResponseHeaders().getFirst("Location")).contains(expected);
    }

    @Test
    void githubCallbackWithoutNonceCookieRedirectsTo400() {
        EntityExchangeResult<byte[]> result = noRedirectClient().get()
                .uri("/auth/github/callback?code=abc&state=java%7Cnonce123")
                .exchange()
                .expectBody()
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(302);
        assertThat(result.getResponseHeaders().getFirst("Location")).endsWith("/error?code=400");
    }

    @Test
    void googleCallbackWithMismatchedNonceRedirectsTo400AndClearsCookie() {
        EntityExchangeResult<byte[]> result = noRedirectClient().get()
                .uri("/auth/google/callback?code=abc&state=java%7Cwrong")
                .header("Cookie", "oauth_state=right")
                .exchange()
                .expectBody()
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(302);
        assertThat(result.getResponseHeaders().getFirst("Location")).endsWith("/error?code=400");
        assertThat(result.getResponseHeaders().get("Set-Cookie"))
                .anyMatch(c -> c.startsWith("oauth_state=;") && c.contains("Max-Age=0"));
    }

    @Test
    void invitationReturnsInvitationWhenValidToken() {
        var inv = TestHelper.createInvitation("test-token-xyz", 2);

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/auth/invitation/" + inv.token())
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
        assertThat(result.getResponseBody()).containsEntry("token", inv.token());
        assertThat(result.getResponseBody()).containsEntry("found", true);
    }
}

File: backends/kotlin-ktor/src/test/kotlin/com/authorization/integration/AuthIntegrationTest.kt
package com.authorization.integration

import com.authorization.module
import com.authorization.handler.signStaffId
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.testing.*
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.serialization.json.*

class AuthIntegrationTest {

    @BeforeTest
    fun setUp() {
        TestHelper.db
        TestHelper.truncateTables()
    }

    @Test
    fun `GET api auth me returns profile when authenticated`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff()
        val response = client.get("/api/auth/me") {
            header(HttpHeaders.Cookie, "staff_id=${TestHelper.signStaffCookie(staff.id)}")
        }
        assertEquals(HttpStatusCode.OK, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertEquals(staff.id, body["staff_id"]!!.jsonPrimitive.long)
    }

    @Test
    fun `GET api auth me returns 401 when unauthenticated`() = testApplication {
        application { module(TestHelper.cfg) }
        val response = client.get("/api/auth/me")
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth me unsigned forged cookie returns 401`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff(email = "forge-1@example.com")
        // 署名を付けず staff_id をそのまま設定した「偽造」クッキー。
        val response = client.get("/api/auth/me") {
            header(HttpHeaders.Cookie, "staff_id=${staff.id}")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth me tampered signature returns 401`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff(email = "forge-2@example.com")
        val signed = TestHelper.signStaffCookie(staff.id)
        // 末尾の1文字を必ず異なる値に置き換える（元の値と偶然一致すると署名が
        // 変わらずテストが不安定になるため）。
        val replacement = if (signed.last() == '0') '1' else '0'
        val tampered = signed.dropLast(1) + replacement
        val response = client.get("/api/auth/me") {
            header(HttpHeaders.Cookie, "staff_id=$tampered")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth me signed with another secret returns 401`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff(email = "forge-3@example.com")
        val forged = signStaffId(staff.id, "attacker-controlled-secret", 3600)
        val response = client.get("/api/auth/me") {
            header(HttpHeaders.Cookie, "staff_id=$forged")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth me expired signed cookie returns 401`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff(email = "forge-4@example.com")
        // 署名自体は正しいが、Max-Ageが切れた後に手動でCookieヘッダーを
        // 再送した状況を再現する（署名対象に有効期限を含めていないと防げない）。
        val expired = signStaffId(staff.id, TestHelper.cfg.app.staffCookieSecret, -3600)
        val response = client.get("/api/auth/me") {
            header(HttpHeaders.Cookie, "staff_id=$expired")
        }
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth login returns info when authenticated`() = testApplication {
        application { module(TestHelper.cfg) }
        val staff = TestHelper.createStaff()
        val response = client.get("/api/auth/login") {
            header(HttpHeaders.Cookie, "staff_id=${TestHelper.signStaffCookie(staff.id)}")
        }
        assertEquals(HttpStatusCode.OK, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertEquals(staff.id, body["staff_id"]!!.jsonPrimitive.long)
    }

    @Test
    fun `GET api auth login returns 401 when unauthenticated`() = testApplication {
        application { module(TestHelper.cfg) }
        val response = client.get("/api/auth/login")
        assertEquals(HttpStatusCode.Unauthorized, response.status)
    }

    @Test
    fun `GET api auth logout returns 200`() = testApplication {
        application { module(TestHelper.cfg) }
        val response = client.get("/api/auth/logout")
        assertEquals(HttpStatusCode.OK, response.status)
    }

    @Test
    fun `GET api auth github redirect issues nonce cookie and embeds it in state`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = createClient { followRedirects = false }
        val response = c.get("/auth/github/redirect?token=inv-token")
        assertEquals(HttpStatusCode.Found, response.status)
        val setCookie = response.headers.getAll(HttpHeaders.SetCookie)!!.first { it.startsWith("oauth_state=") }
        val nonce = Regex("oauth_state=([0-9a-f]+)").find(setCookie)?.groupValues?.get(1)
        assertNotNull(nonce)
        assertTrue(setCookie.contains("HttpOnly"))
        val location = response.headers[HttpHeaders.Location]!!
        assertTrue(location.contains("state=" + java.net.URLEncoder.encode("${TestHelper.cfg.app.runtime}|$nonce|inv-token", "UTF-8")))
    }

    @Test
    fun `GET api auth github callback without nonce cookie redirects to 400`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = createClient { followRedirects = false }
        val response = c.get("/auth/github/callback?code=abc&state=kotlin%7Cnonce123")
        assertEquals(HttpStatusCode.Found, response.status)
        assertTrue(response.headers[HttpHeaders.Location]!!.endsWith("/error?code=400"))
    }

    @Test
    fun `GET api auth google callback with mismatched nonce redirects to 400 and clears cookie`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = createClient { followRedirects = false }
        val response = c.get("/auth/google/callback?code=abc&state=kotlin%7Cwrong") {
            header(HttpHeaders.Cookie, "oauth_state=right")
        }
        assertEquals(HttpStatusCode.Found, response.status)
        assertTrue(response.headers[HttpHeaders.Location]!!.endsWith("/error?code=400"))
        val cleared = response.headers.getAll(HttpHeaders.SetCookie)!!.first { it.startsWith("oauth_state=") }
        assertTrue(cleared.contains("Max-Age=0"))
    }

    @Test
    fun `GET api auth invitation returns invitation when valid token`() = testApplication {
        application { module(TestHelper.cfg) }
        val inv = TestHelper.createInvitation("test-token-xyz")
        val response = client.get("/api/auth/invitation/${inv.token}")
        assertEquals(HttpStatusCode.OK, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertEquals(inv.token, body["token"]!!.jsonPrimitive.content)
        assertTrue(body["found"]!!.jsonPrimitive.boolean)
    }
}
